lib/bolt/transport/local/shell.rb

```diff
@@ -1,26 +1,179 @@
+            end
+          end
+        end
+
+        def with_tmpscript(script)
+          with_tempdir do |dir|
+            dest = File.join(dir.to_s, File.basename(script))
+            copy_file(script, dest)
+            yield dest, dir
+          end
+        end
+
+        def execute(command, sudoable: true, **options)
+          run_as = options[:run_as] || self.run_as
+          escalate = sudoable && run_as && @user != run_as
+          use_sudo = escalate && @target.options['run-as-command'].nil?
+
+          if options[:interpreter]
+            if command.is_a?(Array)
+              command.unshift(options[:interpreter])
+            else
+              command = [options[:interpreter], command]
+            end
+          end
+
+          command_str = command.is_a?(String) ? command : Shellwords.shelljoin(command)
+
+          if escalate
+            if use_sudo
+              sudo_flags = ["sudo", "-k", "-S", "-u", run_as, "-p", Sudoable.sudo_prompt]
+              sudo_flags += ["-E"] if options[:environment]
+              sudo_str = Shellwords.shelljoin(sudo_flags)
+              command_str = "#{sudo_str} #{command_str}"
+            else
+              run_as_str = Shellwords.shelljoin(@target.options['run-as-command'] + [run_as])
+              command_str = "#{run_as_str} #{command_str}"
+            end
+          end
+
+          command_arr = options[:environment].nil? ? [command_str] : [options[:environment], command_str]
+
+          # Prepare the variables!
           result_output = Bolt::Node::Output.new
-          result_output.stdout << stdout unless stdout.nil?
-          result_output.stderr << stderr unless stderr.nil?
-          result_output.exit_code = rc.exitstatus
+          in_buffer = options[:stdin] || ''
+          # Chunks of this size will be read/written in one iteration
+          chunk_size = 4096
+          index = 0
+
+          begin
+            inp, out, err, t = Open3.popen3(*command_arr)
+            read_streams = { out => String.new,
+                             err => String.new }
+            write_stream = in_buffer.empty? ? [] : [inp]
+
+            until Process.waitpid(t.pid, Process::WNOHANG)
+              timeout = 0.1
+              # See if we can read from out or err, or write to in
+              ready_read, ready_write, = select(read_streams.keys, write_stream, nil, timeout)
+
+              # Read from out and err
+              ready_read&.each do |stream|
+                begin
+                  if use_sudo
+                    buffer = stream.readpartial(chunk_size)
+                    # Split on newlines, including the newline
+                    lines = buffer.split(/(?<=[\n])/)
+                    # handle_sudo will return the line if it is not a sudo prompt or error
+                    lines.map! { |line| handle_sudo(inp, line) }
+                    read_streams[stream] << lines.join("")
```

Comment:
It looks like with the current ssh implementation we don't add the lines if they're part of the sudo prompt

Reply:
This doesn't - `handle sudo` returns an empty string if it consumes the sudo prompt. https://github.com/puppetlabs/bolt/pull/897/files#diff-a96c17ec64c4c2d8d0e546f3f96feeb5R559 verifies that it doesn't print anything besides expected stderr.